cudax/include/cuda/experimental/__device/arch_traits.cuh

```diff
@@ -0,0 +1,401 @@
+
+  __traits.cluster_supported  = false;
+  __traits.redux_intrinisic   = true;
+  __traits.elect_intrinsic    = false;
+  __traits.cp_async_supported = true;
+  __traits.tma_supported      = false;
+
+  return __traits;
+}();
+
+inline constexpr arch_traits_t sm_860_traits = []() constexpr {
+  arch_traits_t __traits{};
+  __traits.compute_capability_major             = 8;
+  __traits.compute_capability_minor             = 6;
+  __traits.compute_capability                   = 86;
+  __traits.max_shared_memory_per_multiprocessor = 100 * 1024;
+  __traits.max_blocks_per_multiprocessor        = 16;
+  __traits.max_warps_per_multiprocessor         = 48;
+  __traits.max_threads_per_multiprocessor =
+    __traits.max_warps_per_multiprocessor * detail::arch_common_traits::warp_size;
+  __traits.reserved_shared_memory_per_block = 1024;
+  __traits.max_shared_memory_per_block_optin =
+    __traits.max_shared_memory_per_multiprocessor - __traits.reserved_shared_memory_per_block;
+
+  __traits.cluster_supported  = false;
+  __traits.redux_intrinisic   = true;
+  __traits.elect_intrinsic    = false;
+  __traits.cp_async_supported = true;
+  __traits.tma_supported      = false;
+
+  return __traits;
+}();
+
+inline constexpr arch_traits_t sm_900_traits = []() constexpr {
+  arch_traits_t __traits{};
+  __traits.compute_capability_major             = 9;
+  __traits.compute_capability_minor             = 0;
+  __traits.compute_capability                   = 90;
+  __traits.max_shared_memory_per_multiprocessor = 228 * 1024;
+  __traits.max_blocks_per_multiprocessor        = 32;
+  __traits.max_warps_per_multiprocessor         = 64;
+  __traits.max_threads_per_multiprocessor =
+    __traits.max_warps_per_multiprocessor * detail::arch_common_traits::warp_size;
+  __traits.reserved_shared_memory_per_block = 1024;
+  __traits.max_shared_memory_per_block_optin =
+    __traits.max_shared_memory_per_multiprocessor - __traits.reserved_shared_memory_per_block;
+
+  __traits.cluster_supported  = true;
+  __traits.redux_intrinisic   = true;
+  __traits.elect_intrinsic    = true;
+  __traits.cp_async_supported = true;
+  __traits.tma_supported      = true;
+
+  return __traits;
+}();
+
+// TODO Should this be provided outside detail? Just using arch template seems better
+template <unsigned int __SmVersion>
+_CCCL_HOST_DEVICE inline constexpr arch_traits_t arch_traits()
+{
+  if constexpr (__SmVersion == 60)
+  {
+    return detail::sm_600_traits;
+  }
+  else if constexpr (__SmVersion == 70)
+  {
+    return detail::sm_700_traits;
+  }
+  else if constexpr (__SmVersion == 75)
+  {
+    return detail::sm_750_traits;
+  }
+  else if constexpr (__SmVersion == 80)
+  {
+    return detail::sm_800_traits;
+  }
+  else if constexpr (__SmVersion == 86)
+  {
+    return detail::sm_860_traits;
+  }
+  else
+  {
+    static_assert(__SmVersion == 90, "Unknown architecture");
+    return detail::sm_900_traits;
+  }
+}
```

Comment:
Question: why do we need this function? `constexpr arch_traits_t arch_traits(unsigned int __sm_version)` should be enough, since it's `constexpr`. I don't see where you need `__sm_version` to be a constant expression in the implementation.

Reply:
I could see me removing it, the main difference is handling of a wrong architecture. Right now you will get a static assert. If I remove this function and use one for both static and dynamic context then the error in static context will become:
```
note #2703-D: cannot call non-constexpr function "cuda::__4::__throw_cuda_error" (declared at line 65 of /home/pciolkosz/cccl/libcudacxx/lib/cmake/libcudacxx/../../../include/cuda/std/__exception/cuda_error.h)
[build]         __throw_cuda_error(cudaErrorInvalidValue, "Traits requested for an unknown architecture");
```
Its slightly worse, but possibly still fine. I thought the burden of extra function was worth having the static assert, but I could be easily convinced the other way

Reply:
I believe we have some version of `std::is_constant_evaluated()` in libcu++ that you could use to static_assert instead.

Reply:
Ha, I didn't know it exists! Hmm, but I am still not sure how to structure it. It can't be just static_assert(false) unfortunately, I need to make the condition depend on the function argument. But then it complains:
```
error: expression must have a constant value
note #2689-D: the value of parameter "__sm_version" (declared at line 332) cannot be used as a constant
```
Unless I missunderstand how to use it:
```
    if constexpr (cuda::std::is_constant_evaluated())
    {
      static_assert(__sm_version != 90, "Unknown architecture");
    }
    else
    {
      __throw_cuda_error(cudaErrorInvalidValue, "Traits requested for an unknown architecture");
    }
```

Reply:
for `static_assert(false)` before C++23, you just need to make the `false` a dependent expression, which requires workarounds like wrapping it inside a templated lambda, or moving it to a templated function, and calling it from the else branch.

In any case, `is_constant_evaluated` needs to be called from a run-time `if`, otherwise it's always true:
```
 if (cuda::std::is_constant_evaluated())
```

Reply:
But if its a runtime if, then won't whatever is inside the if always be instantiated and trigger the static_assert, even if default switch branch is not reached? That is what I am seeing when I added a templated wrapper for static_assert
```
template <typename T>
void static_assert_function()
{
  static_assert(sizeof(T) == 0);
}
...
  default:
    if (cuda::std::is_constant_evaluated())
    {
      static_assert_function<Dummy>();
    }
```

Reply:
I couldn't make it work, but you convinced me to remove the templated version, error message from the throw is good enough here